source/extensions/tracers/xray/tracer.cc

```diff
@@ -94,6 +95,19 @@ void Span::finishSpan() {
     s.mutable_annotations()->insert({item.first, item.second});
   }
 
+  // Gets this Span's direction from operation name.
+  // `operation_name_` can be either "ingress" or "egress <header_hostname>"
+  // which directly comes from `span_name` that is populated inside function
+  // `HttpTracerImpl::startSpan` at source/common/tracing/http_tracer_impl.cc
+  // The `substr` operation below will extract only the first word from
+  // `operation_name_`, so the `direction` is either "ingress" or "egress".
+  const size_t pos = operation_name_.find(' ');
+  if (pos == std::string::npos) {
+    s.mutable_annotations()->insert({DirectionKey, operation_name_});
+  } else {
+    s.mutable_annotations()->insert({DirectionKey, operation_name_.substr(0, pos)});
+  }
```

Comment:
String type `operation_name_` is not a good choice to determine the direction of traffic. Because it can be simply overwritten by some other modules, such as filters that will operate tracing span, **decorators of request/response**, etc.

The `OperationName operationName()` of Tracing::Config would be the better. Just like Zipkin did:

```
SpanPtr Tracer::startSpan(const Tracing::Config& config, const std::string& span_name,
                          SystemTime timestamp) {
  // Build the endpoint
  Endpoint ep(service_name_, address_);

  // Build the CS annotation
  Annotation cs;
  cs.setEndpoint(std::move(ep));
  if (config.operationName() == Tracing::OperationName::Egress) {
    cs.setValue(CLIENT_SEND);
  } else {
    cs.setValue(SERVER_RECV);
  }

  // ......

```

Reply:
Thanks for the review. I was debating whether to use the `config.operationName()` as I would have to modify the function signature at few places. Now that you mention that `operation_name_` is not a good choice, I see that it will not always come from `startSpan`, but also set from `spawnChild` from several places.

I have modified the implementation to read the direction from `config.operationName()` directly.